zerver/actions/message_send.py

```diff
@@ -858,11 +861,59 @@ def get_active_presence_idle_user_ids(
     return filter_presence_idle_user_ids(user_ids)
 
 
-@transaction.atomic(savepoint=False)
+def insert_idempotent_msg(message: Message, idempotency_key: str) -> SentMessageResult | None:
```

Comment:
`insert_message_with_idempotency_key` would be a better name. `msg` is an abbreviation we avoid across the codebase, and this is a lot more clear.